src/Npgsql/TypeMapping/ConnectorTypeMapper.cs

```diff
@@ -511,6 +511,11 @@ public override void AddTypeResolverFactory(ITypeHandlerResolverFactory resolver
 
                 if (Interlocked.CompareExchange(ref _resolvers, newResolvers, oldResolvers) == oldResolvers)
                 {
+                    _handlersByOID.Clear();
```

Comment:
I think this clearing should be part of the lock, technically you could get init races (or partial inits) again

Reply:
That's true, but this would be a race condition between a Reset and a ResolveBy method, right? If so, do you see a particular problem?

More generally, do you see an issue with the lack of write locking? It's true that we could end up with two identical type handler instances - when for resolving by OID, another by NpgsqlDbType, but is that really something worth protecting against?

Reply:
(note there's no write locking at all any more!)

Reply:
Decided to add back write locking in an offline conversation, to be a bit more future proof (mostly there's no downside).